src/utilities/sorting/sorting.ts

```diff
@@ -0,0 +1,70 @@
+import {CARTA} from "carta-protobuf";
+
+import {ControlHeader} from "stores";
+import {ProcessedColumnData} from "utilities";
+
+export function updateSortedIndexMapFunc(
+    controlHeader: Map<string, ControlHeader>,
+    sortingInfo: {columnName: string; sortingType: CARTA.SortingType},
+    sortedIndexMap: Array<number>,
+    hasFilter: boolean,
+    numVisibleRows: number,
+    sortData: Map<number, ProcessedColumnData>
+) {
+    const dataIndex = controlHeader.get(sortingInfo.columnName)?.dataIndex;
+
+    if (dataIndex >= 0) {
+        let direction = 0;
+
+        if (sortingInfo.sortingType != null) {
+            direction = sortingInfo.sortingType ? -1 : 1;
+        } else {
+            return initSortedIndexMapFunc(numVisibleRows);
+        }
+
+        if (hasFilter) {
+            return initSortedIndexMapFunc(numVisibleRows);
+        }
+        let queryColumn = sortData.get(dataIndex);
+
+        switch (queryColumn?.dataType) {
+            case CARTA.ColumnType.String:
+                sortedIndexMap.sort((a: number, b: number) => {
+                    const aString = String(queryColumn.data[a]);
+                    const bString = String(queryColumn.data[b]);
+                    if (!aString) {
+                        return direction * -1;
+                    }
+
+                    if (!bString) {
+                        return direction * 1;
+                    }
+                    return direction * aString.localeCompare(bString);
+                });
+                break;
+            case CARTA.ColumnType.UnsupportedType:
+                console.log("Data type is not supported");
+                break;
+            default:
+                sortedIndexMap.sort((a: number, b: number) => {
+                    const aNumber = Number(queryColumn.data[a]);
+                    const bNumber = Number(queryColumn.data[b]);
+                    return direction * (aNumber < bNumber ? -1 : 1);
+                });
+                break;
+        }
+    }
+    return sortedIndexMap;
+}
+
+export function initSortedIndexMapFunc(numVisibleRows: number) {
```

Comment:
I suggest we use a more general name `getInitIndexMap` and also apply it to `CatalogOnlineQueryProfileStore.initFilterIndexMap`.

Reply:
Could we apply it to `CatalogOnlineQueryProfileStore.initFilterIndexMap`?

Reply:
Yes, we can.